### PR TITLE
Streaming ASR guide/tutorial

### PR DESCRIPTION
**Pull request by asumagic**
## What does this PR do?

This PR adds a new, full guide on:

- Architectural changes to a Conformer model for chunkwise streaming
  - Includes explanations on dynamic chunk training, dynamic chunk convolutions, chunked attention
- Providing an example of some of the streaming-specific debugging tools in SpeechBrain
- Changes to train a streaming capable model in SpeechBrain
- Making a trained streaming model ready for inference 
- Demonstrating StreamingASR capabilities

Complete with illustrations, a healthy amount of documentation/code links and code snippets.

It is added under the neural architectures section of tutorials.

**PR preview:** https://speechbrain--2700.org.readthedocs.build/en/2700/tutorials/nn/conformer-streaming-asr.html

It also adds an `assets/` directory inside of `docs/tutorials/`, because it turns out Colab handles relative image links in GitHub repos seemingly just fine, so there is no need to base64 the images.  
Might not _retroactively_ do this, because it would just add to the repo size currently...

<details>
  <summary><b>Before submitting</b></summary>

- [x] Did you read the [contributor guideline](https://speechbrain.readthedocs.io/en/latest/contributing.html)?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing [tests](https://github.com/speechbrain/speechbrain/tree/develop/tests) pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Does your code adhere to project-specific code style and conventions?

</details>

## PR review

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified
- [ ] Confirm that the changes adhere to compatibility requirements (e.g., Python version, platform)
- [ ] Review the self-review checklist to ensure the code is ready for review

</details>

### CODE REVIEW

**Review by TParcollet — COMMENTED** [review]
1. conformer-simple.png is confusing for RNN-T as you omitted the Transducer predictor network. Maybe just adding a single block without details would help. 
2. The relative pos part is a bit confusing, maybe lack a few details to understand the figure. 
3. Section on losses: cross entropy training with encoder-decoder is not necessarily a problem tbh. It can be fixed. Maybe rewrite saying that it's not the easiest / best way to do it that is available or smth like that. 
4. "Currently, only greedy search is supported in a streaming context. You probably want to make it so that your test set is evaluated with greedy search." -- Is that actually true? 
5. Should we have a section TL;DR summarizing briefly what needs to be known?

**Comment by TParcollet** [comment]
Bruh dropped a lecture on streaming conformers.

**Comment by asumagic** [comment]
> conformer-simple.png is confusing for RNN-T as you omitted the Transducer predictor network. Maybe just adding a single block without details would help.

Will do.

> The relative pos part is a bit confusing, maybe lack a few details to understand the figure.

Will try.

> Section on losses: cross entropy training with encoder-decoder is not necessarily a problem tbh. It can be fixed. Maybe rewrite saying that it's not the easiest / best way to do it that is available or smth like that.

That's probably true. I haven't actually looked at it in detail but in retrospect I think it's probably feasible. I'll edit up the wording accordingly.

(Unrelated: Even without that I wonder if training a RNN-T model with joint CTC and CE AED losses would be beneficial to model accuracy even if the attentive decoder doesn't end up being used in RNN-T. IIRC the dynamic chunk convolution paper has done that.)

> "Currently, only greedy search is supported in a streaming context. You probably want to make it so that your test set is evaluated with greedy search." -- Is that actually true?

I actually haven't bothered to try CTC. It probably needs a simple no-op wrapper for decoding...  
But for RNN-T, this is true. I have only ever implemented streaming GS.

> Should we have a section TL;DR summarizing briefly what needs to be known?

I tried to avoid it by making splitting into subsection, and splitting sections into theory and practice as much as possible, but it still ended up being so full of content I might as well make a summary... Will do.

**Comment by asumagic** [comment]
Had to reread stuff a bit but see the commits and my comment above for updates @TParcollet. (Note that the docs might still be building.)

Make sure I haven't added error especially in the more theoretical part of positional embeddings as a bit of the `RelPosMHAXL` math remains obscure to me...

**Comment by TParcollet** [comment]
@asumagic thanks, if the tests pass, i'll merge.

**Review by TParcollet — APPROVED** [review]
With the changes, it LGTM.
